articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
 
 ----
 
-== Parameter usage prevents Cypher injection
+The query that will be run is:
 
-We can easily prevent Cypher injection by always submitting input as parameters to our query.
 
-https://neo4j.com/docs/cypher-manual/current/syntax/parameters/
+[source,cypher]
+----
+CREATE (s:Student) SET s.name = 'Robby' WITH DISTINCT true as haxxored MATCH (s:Student) DETACH DELETE s //';
+----
+
+After the Robby node is created, all :Student nodes are found and deleted.
+
+== Types of Injection Attacks
+
+In the previous example we showed how Little Robby could ruin your day by deleting all the data in the database.
+But this is not the only way a malicious actor can exploit queries by injection.
+
+=== Information disclosure
+
+A possible injection vector is when the malicious input aims to read information that they should not have access to.
+
+For example; the attack payload of:
+
+`Robby' OR 1=1 RETURN apoc.text.join(collect(n.username), ',') //`
+
+Might execute as;
+
+[source,cypher]
+----
+MATCH (s:Student) WHERE s.name = 'Robby' OR 1=1 RETURN apoc.text.join(collect(n.username), ',') //' RETURN s.name;
+----
+
+Returning the names of all the users in the database in one comma separated string.
+For this method to succeed, the client application needs to be vulnerable to injection as well as sending back the results of the query to the user.
+
+=== Blind Injection
+
+Blind injection is when the attacker doesn't aim to fetch the disclosed information from the client response directly, but may obtain it in a different way.
+
+One way this is done is by reacting to the behavior of the application. Let's assume that a website loads a different page based on the existence result of a query.
+For example a login page first asks for an email and will then display either a login to continue page or a register to continue page.
+
+`MATCH (u) WHERE user.email = '{email}' RETURN u IS NOT NULL;`
```

Comment:
The variables used MUST be consistent. This mixes up `u` and `user` variables, such that this example will error out and cannot be run.